tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/package.scala

```diff
@@ -84,4 +87,32 @@ package object transformations {
    * A transformation that makes a module out of a module.
    */
   type TlaModuleTransformation = TlaModule => TlaModule
+
+  /**
+   * Returns a transformation which primes all unprimed variables from `vars` appearing in a given expression.
+   *
+   * Example:
+   *
+   * {{{
+   * a' + b > 0 --> a' + b' > 0
+   * }}}
+   *
+   * @param vars
+   *   variables that should be primed
+   * @param tracker
+   *   a transformation tracker for recording which expressions have changed
+   * @return
+   *   the transformation that adds primes to the variables from `vars`
+   */
+  def decorateWithPrime(vars: Set[String], tracker: TransformationTracker): TlaExTransformation = {
+    ReplaceFixed(tracker).withFun {
+      case ex @ NameEx(name) if vars.contains(name) =>
+        // add prime to a variable from vars
+        OperEx(TlaActionOper.prime, ex)(ex.typeTag)
+
+      case OperEx(TlaActionOper.prime, primedOnce @ OperEx(TlaActionOper.prime, _)) =>
+        // in case the variable was already primed, remove the second prime
+        primedOnce
```

Comment:
Rather than defining de-priming of doubly primed expressions, which relies on the implementation detail that `ReplaceFixed` implements post-transform too 
https://github.com/informalsystems/apalache/blob/919e65948904156615a05980d98b335fd5a8fc2a/tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/ReplaceFixed.scala#L96-L99
I'd prefer defining this in a way that never attempts to add unnecessary primes in the first place:
```suggestion
      case ex@OperEx(TlaActionOper.prime, _) =>
        // in case the expression was already primed, ignore
        ex
        
      case ex @ NameEx(name) if vars.contains(name) =>
        // add prime to a variable from vars
        OperEx(TlaActionOper.prime, ex)(ex.typeTag)
```

Reply:
Are you sure that this code would not introduce two primes, i.e., by going to 113-115 first and then to 109-111 one level up in the recursion?

Reply:
When it goes one level up it will already be primed, so 109 will skip.
```
x -(113)-> x' -(109)-> x'
x' -(109)-> x' -(109)-> x'
```

Reply:
Ah, that makes sense.

Reply:
I looked into it, seems like your implementation of `withFun` doesn't allow short-circuiting.
Unlike a dedicated transformation, like `Prime`, it must always first look into the generic `OperEx` case
https://github.com/informalsystems/apalache/blob/919e65948904156615a05980d98b335fd5a8fc2a/tlair/src/main/scala/at/forsyte/apalache/tla/lir/transformations/standard/ReplaceFixed.scala#L96-L99
This is potentially problematic in practice; a lot of transformations rely on the fact that certain synactic forms halt exploration, because transforming may ruin subexpressions, but these generic `PartialFunction` traversals do not, they always full explore the expression tree.

Reply:
True. `Prime` had a very special behavior. We can think of generalizing `ReplaceFixed` even further.

Reply:
It's not just Prime, a typical tx has cases like 
```scala
case InterestingCase => interestingCaseTx
case OperEx(args) => mapOverArgs
case LetIn => mapOverDecls
case default => defaultTx
```

where `interestingCaseTx` would typically not recurse any more. I think this can be fixed by changing
```scala
    ex match {
      case LetInEx(body, defs @ _*) =>
        // Transform bodies of all op.defs
        val newDefs = defs.map(tracker.trackOperDecl { d => d.copy(body = self(d.body)) })
        val newBody = self(body)
        val retEx = if (defs == newDefs && body == newBody) ex else LetInEx(newBody, newDefs: _*)(ex.typeTag)
        tr(retEx)


      case OperEx(op, args @ _*) =>
        val newArgs = args.map(self)
        val retEx = if (args == newArgs) ex else OperEx(op, newArgs: _*)(ex.typeTag)
        tr(retEx)


      case _ => tr(ex)
```
in `withFun` to
```scala
    ex match {
      case special if partialFun.isDefinedAt(special) =>
        partialFun(special)

      case LetInEx(body, defs @ _*) =>
        // Transform bodies of all op.defs
        val newDefs = defs.map(tracker.trackOperDecl { d => d.copy(body = self(d.body)) })
        val newBody = self(body)
        val retEx = if (defs == newDefs && body == newBody) ex else LetInEx(newBody, newDefs: _*)(ex.typeTag)
        retEx

      case OperEx(op, args @ _*) =>
        val newArgs = args.map(self)
        val retEx = if (args == newArgs) ex else OperEx(op, newArgs: _*)(ex.typeTag)
        retEx

      case _ => ex
```

Reply:
It got automerged. Shall we open another PR for your suggestions?